### ISSUE TITLE
about self.gamma

### ISSUE THREAD

**Issue by fenghao195** [Jungjaewon/Single_Image_SuperResolution_via_Holistic_Attention_Network#2]
Hi,I find that self.gamma is not trained，and requires_grad of tensor is True.model do not contain it

**Comment by Jungjaewon** [comment]
Thanks for you interest of my repo. I cannot find self.gamma in the the code. Could you tell me where is self.gamma?.

**Comment by fenghao195** [comment]
Sorry，I find that self.scale of Layer Attention is not trained，and requires_grad of tensor is True.model do not contain it

**Comment by Jungjaewon** [comment]
I've found pytorch document. The default of require_grad of nn.Parameter is True. Then, the parameter is trainable as the same as in Channel_Spatial_Attention_Module.

**Comment by fenghao195** [comment]
![image](https://user-images.githubusercontent.com/76979950/191634236-88133c1e-61b4-4804-bf79-d65a0a36e06a.png)
![image](https://user-images.githubusercontent.com/76979950/191634267-54dd8270-a284-4537-ad53-886467d0f60f.png)
Look,The default of require_grad of nn.Parameter is True,but print of model do not contain

**Comment by fenghao195** [comment]
![image](https://user-images.githubusercontent.com/76979950/191634541-5b40ec83-1b24-4f7d-b5eb-c9c2977cafc2.png)
LAM_Module  block only contain a softmax function

**Comment by fenghao195** [comment]
I haved consulted pytorch doc,I think there is nothing wrong with the way I use it。but,value of self.scale is changeless. You can try

**Comment by Jungjaewon** [comment]
OH... I see..., I will put require_grad=True on nn.Parameters for both modules. Really thanks for your correction.

**Comment by fenghao195** [comment]
................
I don't think you wrote it wrong. I just don't know why this value can't be trained.ps:(I know its  require_grad=True,But it really can't be trained，you can try) My expression ability is limited

**Comment by Jungjaewon** [comment]
Okay for now, I am on different task and figure out this things ASAP.  I have a question. what the version of pytorch do you use for the testing?

**Comment by fenghao195** [comment]
pytorch:torch1.8.0+cu111,python:3.6

**Comment by Jungjaewon** [comment]
Sorry for being late. 

I wrote code for testing to check nn.paramter(,requires_grad=True) works or not.

In short, it works, however, we cannot see the parameter when we print a model. It is a bug of PyTorch?

`
import torch.nn as nn
import torch
import random

from torch.utils import data

class ExampleDataset(data.Dataset):

    def __init__(self, num_data=5000):
        self.data_list = list()
        for _ in range(num_data):
            self.data_list.append(random.randint(0, 10000))

    def __getitem__(self, index):
        in_d = self.data_list[index]
        return in_d, 2 * in_d

    def __len__(self):
        """Return the number of images."""
        return len(self.data_list)


class Test_Module(nn.Module):
    def __init__(self, rq=False):
        super(Test_Module, self).__init__()
        self.a = nn.Parameter(torch.zeros(1), requires_grad=rq)
        self.softmax = nn.Softmax(dim=2)

        print(f'scale : {self.a}')

    def forward(self, x):
        return self.a * x


def train_example(rq=True):
    print(f'rq : {rq}')
    model = Test_Module(rq=rq)
    print(model)

    l1_loss = nn.L1Loss()
    dataset = ExampleDataset()
    optimizer = torch.optim.Adam(model.parameters(), 0.1, (0.9, 0.999))

    data_loader = data.DataLoader(dataset=dataset,
                                  batch_size=1,
                                  shuffle=True,
                                  num_workers=1,
                                  drop_last=True)
    for x, y in data_loader:
        # print(x,y)
        pred = model(x)
        loss = l1_loss(pred, y)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        print(f'loss : {loss}, a : {model.a}, input : {x}, pred : {pred}')

if __name__ == '__main__':
    train_example(rq=True)
    train_example(rq=False)
`

When rq is False, we cannot train a model. 

Thanks for your interest of my repository

**Comment by Jungjaewon** [comment]
<img width="819" alt="Screen Shot 2022-10-09 at 8 14 33 PM" src="https://user-images.githubusercontent.com/7018886/194753777-0b3f8fed-931e-403e-81e2-157a43ee0dfb.png">

My environment is pytorch1.7.1 cuda11.2 python3.7.13.
You should run my code

**Comment by fenghao195** [comment]
ok,thank you very much!!!


emmm!
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Jungjaewon/Single_Image_SuperResolution_via_Holistic_Attention_Network"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月9日(星期天) 晚上7:14
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Jungjaewon/Single_Image_SuperResolution_via_Holistic_Attention_Network] about self.gamma (Issue #2)





 
Sorry for being late.
 
I wrote code for testing to check nn.paramter(,requires_grad=True) works or not.
 
In short, it works, however, we cannot see the parameter when we print a model. It is a bug of pytorch?.
 
`
 import torch.nn as nn
 import torch
 import random
 
from torch.utils import data
 
class ExampleDataset(data.Dataset):
 def __init__(self, num_data=5000):     self.data_list = list()     for _ in range(num_data):         self.data_list.append(random.randint(0, 10000)) def __getitem__(self, index):     in_d = self.data_list[index]     return in_d, 2 * in_d def __len__(self):     """Return the number of images."""     return len(self.data_list)  
class Test_Module(nn.Module):
 def init(self, rq=False):
 super(Test_Module, self).init()
 self.a = nn.Parameter(torch.zeros(1), requires_grad=rq)
 self.softmax = nn.Softmax(dim=2)
     print(f'scale : {self.a}') def forward(self, x):     return self.a * x  
def train_example(rq=True):
 print(f'rq : {rq}')
 model = Test_Module(rq=rq)
 print(model)
 l1_loss = nn.L1Loss() dataset = ExampleDataset() optimizer = torch.optim.Adam(model.parameters(), 0.1, (0.9, 0.999)) data_loader = data.DataLoader(dataset=dataset,                               batch_size=1,                               shuffle=True,                               num_workers=1,                               drop_last=True) for x, y in data_loader:     # print(x,y)     pred = model(x)     loss = l1_loss(pred, y)     optimizer.zero_grad()     loss.backward()     optimizer.step()     print(f'loss : {loss}, a : {model.a}, input : {x}, pred : {pred}')  
if name == 'main':
 train_example(rq=True)
 train_example(rq=False)
 `
 
When rq is False, we cannot train a model.
 
Thanks for you interest of my respository

**Comment by fenghao195** [comment]
OK，Thank you very much！！！


emmm!
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Jungjaewon/Single_Image_SuperResolution_via_Holistic_Attention_Network"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2022年10月9日(星期天) 晚上7:15
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Jungjaewon/Single_Image_SuperResolution_via_Holistic_Attention_Network] about self.gamma (Issue #2)





 

 
My environment is pytorch1.7.1 cuda11.2 python3.7.13.
 You should run my code
